ProjFS.Mac/PrjFSKextTests/VnodeCacheTests.mm

```diff
@@ -0,0 +1,700 @@
+    // set the entry in the cache to RootHandle_Indeterminate
+    VirtualizationRootHandle rootHandle = VnodeCache_InvalidateVnodeAndGetLatestRoot(
+        &self->dummyPerfTracer,
+        PrjFSPerfCounter_VnodeOp_Vnode_Cache_Hit,
+        PrjFSPerfCounter_VnodeOp_Vnode_Cache_Miss,
+        PrjFSPerfCounter_VnodeOp_FindRoot,
+        PrjFSPerfCounter_VnodeOp_FindRoot_Iteration,
+        self->testVnodeFile1.get(),
+        self->dummyVFSContext);
+    XCTAssertTrue(rootHandle == repoRootHandle);
+    XCTAssertTrue(testVnodeVid == cacheWrapper[indexFromHash].vid);
+    XCTAssertTrue(RootHandle_Indeterminate == cacheWrapper[indexFromHash].virtualizationRoot);
+    
+    XCTAssertFalse(MockCalls::DidCallAnyFunctions());
+}
+
+- (void)testVnodeCache_InvalidateCache_SetsMemoryToZeros {
+    VnodeCacheEntriesWrapper cacheWrapper(/* capacity*/ 100, /* fillCache*/ true);
+
+    shared_ptr<VnodeCacheEntry> emptyArray(static_cast<VnodeCacheEntry*>(calloc(cacheWrapper.GetCapacity(), sizeof(VnodeCacheEntry))), free);
+    XCTAssertTrue(0 != memcmp(emptyArray.get(), s_entries, sizeof(VnodeCacheEntry) * cacheWrapper.GetCapacity()));
+    
+    VnodeCache_InvalidateCache(&self->dummyPerfTracer);
+    XCTAssertTrue(0 == memcmp(emptyArray.get(), s_entries, sizeof(VnodeCacheEntry) * cacheWrapper.GetCapacity()));
+    
+    XCTAssertFalse(MockCalls::DidCallAnyFunctions());
+}
+
+- (void)testInvalidateCache_ExclusiveLocked_SetsMemoryToZeros {
+    VnodeCacheEntriesWrapper cacheWrapper(/* capacity*/ 100, /* fillCache*/ true);
+    
+    shared_ptr<VnodeCacheEntry> emptyArray(static_cast<VnodeCacheEntry*>(calloc(cacheWrapper.GetCapacity(), sizeof(VnodeCacheEntry))), free);
+    XCTAssertTrue(0 != memcmp(emptyArray.get(), s_entries, sizeof(VnodeCacheEntry) * cacheWrapper.GetCapacity()));
+    
+    InvalidateCache_ExclusiveLocked();
+    XCTAssertTrue(0 == memcmp(emptyArray.get(), s_entries, sizeof(VnodeCacheEntry) * cacheWrapper.GetCapacity()));
+    
+    XCTAssertFalse(MockCalls::DidCallAnyFunctions());
+}
+
+- (void)testComputePow2CacheCapacity {
+
+    // At a minimum ComputePow2CacheCapacity should return the minimum value in AllowedPow2CacheCapacities
+    XCTAssertTrue(MinPow2VnodeCacheCapacity == ComputePow2CacheCapacity(0));
+    
+    // ComputePow2CacheCapacity should round up to the nearest power of 2 (after multiplying expectedVnodeCount by 2)
+    int expectedVnodeCount = MinPow2VnodeCacheCapacity/2 + 1;
+    XCTAssertTrue(MinPow2VnodeCacheCapacity << 1 == ComputePow2CacheCapacity(expectedVnodeCount));
+    
+    // ComputePow2CacheCapacity should be capped at the maximum value in AllowedPow2CacheCapacities
+    XCTAssertTrue(MaxPow2VnodeCacheCapacity == ComputePow2CacheCapacity(MaxPow2VnodeCacheCapacity + 1));
+    
+    XCTAssertFalse(MockCalls::DidCallAnyFunctions());
+}
+
+- (void)testComputeVnodeHashKeyWithCapacityOfOne {
+    VnodeCacheEntriesWrapper cacheWrapper(/* capacity*/ 1, /* fillCache*/ false);
```

Comment:
Out of curiosity how many tests did you have to add to get 100% coverage.  Tests you wouldn't have added otherwise.

Were you able to find bugs with the unit tests?

Reply:
The coverage validation helped me make sure I was covering all of the cases in `FindVnodeRootFromDiskAndUpdateCache` (I'm not sure that I would have hit all of these without it).

The unit tests did not reveal any bugs.